presto-main/src/main/java/io/prestosql/operator/SimplePageWithPositionEqualsAndHash.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.type.Type;
+import io.prestosql.type.BlockTypeOperators;
+import io.prestosql.type.BlockTypeOperators.BlockPositionHashCode;
+import io.prestosql.type.BlockTypeOperators.BlockPositionIsDistinctFrom;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class SimplePageWithPositionEqualsAndHash
+        implements PageWithPositionEqualsAndHash
+{
+    private final IntList equalityChannels;
+    private final List<BlockPositionIsDistinctFrom> distinctFromOperators;
+    private final List<BlockPositionHashCode> hashOperators;
+
+    public SimplePageWithPositionEqualsAndHash(List<Type> channelTypes, List<Integer> equalityChannels, BlockTypeOperators blockTypeOperators)
+    {
+        requireNonNull(channelTypes, "channelTypes is null");
+        this.equalityChannels = new IntArrayList(requireNonNull(equalityChannels, "equalityChannels is null"));
+        checkArgument(channelTypes.size() >= equalityChannels.size(), "channelTypes cannot have fewer columns then equalityChannels");
+
+        // Use IS DISTINCT FROM for equality, because it evaluates NULL values as distinct (unlike SQL EQUALS).
```

Comment:
`Double.NaN` are also not distinct from each other. Did you check standard how `rank()` function should be implemented? Or is it DB implementation detail?

Reply:
NaN behavior is actually not specified in SQL. @dain has a series of PR to try to modify it however due to some user suggestions, but apparently unspecified in spec. The policy here is just to go with whatever we do from grouping.